### PR TITLE
fix(claude/setup): #589 canonicalize DOTFILES_ROOT to main worktree

### LINKED ISSUE #589: fix(claude/setup): 워크트리에서 setup 실행 시 ~/.claude-*/ 심볼릭이 워크트리 경로로 박혀 dangling 발생
## TL;DR

`./setup.sh` (또는 셸 시작 시 `claude_accounts_init`) 가 **워크트리 디렉토리에서** 호출되면 `$DOTFILES_ROOT` 가 그 워크트리 경로로 평가됨. 그 경로 기반으로 `~/.claude-<acct>/{settings.json, statusline-command.sh, settings.local.json}` 심볼릭이 생성된다. 이후 워크트리가 삭제되면 심볼릭이 dangling 되어 **Claude Code 가 사용자 settings.json 을 읽지 못하고 default 만 로드** → statusline 미동작 + hooks 미실행 + enabledPlugins 누락 등 silent regression.

## 증상 (Symptom)

- statusline 이 표시되지 않음 (`statusLine.command` 이 default 에 없음 + 설령 있어도 dangling 타겟이라 ENOENT)
- `Stop` / `PostToolUse` hooks 미동작
- `enabledPlugins`, `skipDangerousModePermissionPrompt`, `model` 등 settings.json 키가 전부 default 로 동작
- 사용자 입장에서는 \"왜 안 되지\" — 에러도 안 나옴 (broken symlink 는 stat 시 ENOENT, Claude Code 는 settings.json 부재 시 default 로 진행)

## 재현 (Reproduction)

실제 관찰 (2026-05-13, 본 PC):

```sh
$ env | grep CLAUDE_CONFIG_DIR
CLAUDE_CONFIG_DIR=/home/bwyoon/.claude-work1

$ ls -la ~/.claude-work1/settings.json ~/.claude-work1/statusline-command.sh
lrwxrwxrwx ... settings.json         -> /home/bwyoon/dotfiles-bug-2/claude/settings.json
lrwxrwxrwx ... statusline-command.sh -> /home/bwyoon/dotfiles-bug-2/claude/statusline-command.sh

$ ls /home/bwyoon/dotfiles-bug-2
ls: cannot access '/home/bwyoon/dotfiles-bug-2': No such file or directory

$ git -C ~/dotfiles worktree list
/home/bwyoon/dotfiles              [main]
/home/bwyoon/dotfiles-bug-3        [wt/bug/3]
/home/bwyoon/dotfiles-issue-575-1  [wt/issue-575/1]
# dotfiles-bug-2 는 이미 삭제됨

$ test -e ~/.claude-work1/settings.json && echo OK || echo BROKEN
BROKEN

$ stat -c '%y' ~/.claude-work1/settings.json ~/.claude-work1/statusline-command.sh \
                ~/.claude-work1/settings.local.json
2026-05-12 17:33:xx   ← 3 개 모두 동일 시각 (단일 setup 실행에서 생성)
```

3 개 심볼릭의 mtime 이 같다는 점이 \"같은 setup 호출에서 한꺼번에 만들어졌다\" 는 증거.

다른 PC 에서 의도적으로 재현하려면:

```sh
git worktree add ../dotfiles-tmp some-branch
cd ../dotfiles-tmp
./setup.sh                              # ← $DOTFILES_ROOT 가 워크트리로 평가됨
cd ..
git worktree remove dotfiles-tmp        # 또는 디렉토리 직접 rm -rf
ls -la ~/.claude-*/settings.json        # → 워크트리 경로를 가리키는 dangling 심볼릭
```

## 근본 원인 (Root Cause)

`shell-common/tools/integrations/claude.sh` 의 모든 symlink 생성 호출이 `${DOTFILES_ROOT}` 를 진입 시점에 evaluate 된 값으로 사용한다:

- `_claude_account_setup_one` (claude.sh:1027-1049):
  ```sh
  _claude_ensure_symlink "${DOTFILES_ROOT}/claude/settings.json"          "$_caso_cdir/settings.json"
  _claude_ensure_symlink "${DOTFILES_ROOT}/claude/statusline-command.sh"  "$_caso_cdir/statusline-command.sh"
  _claude_ensure_symlink "${DOTFILES_ROOT}/claude/global-memory"          "$_caso_cdir/projects/GLOBAL/memory"
  ```
- `_claude_dir_sync_one` (claude.sh:862): `_cdso_src_root="${DOTFILES_ROOT:-$HOME/dotfiles}/claude/$_cdso_name"`

`$DOTFILES_ROOT` 는 `shell-common/env/_paths.sh` 류에서 `cd` 기준 또는 현재 셸의 실행 디렉토리 기준으로 결정되므로, 워크트리 안에서 setup/init 가 실행되면 그 워크트리 경로가 캡처된다. **메인 워크트리 경로로 정규화하는 가드가 없다.**

ai-worktree-spawn 시나리오에서 특히 위험:
1. 사용자가 메인에서 `ai-worktree:spawn` 으로 워크트리 생성 → 그 안으로 cd
2. 워크트리 안에서 셸 재시작 (또는 새 터미널 진입) → `claude_accounts_init` 자동 호출 → `$DOTFILES_ROOT=/path/to/worktree`
3. 심볼릭이 워크트리 경로로 재바인딩됨 (기존 메인 경로 심볼릭을 덮어씀 — `_claude_ensure_symlink` 의 \"stale 심볼릭 → 교체\" 로직)
4. 워크트리 작업 종료 + `ai-worktree:teardown` 으로 워크트리 삭제 → dangling

`_claude_ensure_symlink` (claude.sh:760-791) 는 기존 심볼릭의 타겟이 \"인자로 받은 src 와 동일한가\" 만 검사하므로, 메인 → 워크트리 → 다른 워크트리 → ... 로 자유롭게 덮어써진다. 메인 경로를 우대하는 정책 없음.

### Option A — `$DOTFILES_ROOT` 를 메인 워크트리로 정규화 (권장)

`shell-common/env/` 의 paths 초기화 단계에서:

```sh
# 현재 dir 이 워크트리면 메인 워크트리 경로를 쓴다.
_resolve_dotfiles_root() {
    _candidate="${DOTFILES_ROOT:-$HOME/dotfiles}"
    if command -v git >/dev/null 2>&1 && [ -d "$_candidate/.git" ] || [ -f "$_candidate/.git" ]; then
        _common_dir=$(git -C "$_candidate" rev-parse --git-common-dir 2>/dev/null)
        if [ -n "$_common_dir" ]; then
            # .git/worktrees/<name>/  형태면 메인 워크트리로 올라간다
            case "$_common_dir" in
                */.git/worktrees/*) _main_git=$(dirname "$(dirname "$_common_dir")") ;;
                */.git|.git)         _main_git=$(dirname "$_common_dir") ;;
            esac
            [ -d "$_main_git" ] && _candidate="$_main_git"
        fi
    fi
    DOTFILES_ROOT="$_candidate"
    export DOTFILES_ROOT
}
```

**장점**: 단일 진입점 수정. 워크트리에서 무엇을 하든 심볼릭은 메인을 가리킴.
**단점**: 사용자가 의도적으로 워크트리 dotfiles 를 테스트하려는 경우 우회 escape 필요 (`DOTFILES_ROOT_FORCE` 같은 env).

### Option B — `setup.sh` 진입 시 워크트리 가드

```sh
# setup.sh 맨 앞
_git_dir=$(git rev-parse --git-dir 2>/dev/null)
_git_common=$(git rev-parse --git-common-dir 2>/dev/null)
if [ "$_git_dir" != "$_git_common" ]; then
    log_error \"setup.sh 는 메인 워크트리에서만 실행하세요.\"
    log_error \"현재: $(pwd) (워크트리)\"
    log_error \"메인으로 이동: cd $(dirname \"$(dirname \"$_git_common\")\")\"
    exit 1
fi
```

**장점**: 명시적, 사용자에게 가시적.
**단점**: `claude_accounts_init` 가 셸 시작마다 동작하는 경로는 못 막음 (A 가 더 광범위).

### Option C — `claude_accounts_init` 에 dangling-symlink 자동 복구

매 셸 시작마다 `~/.claude-*/settings.json` 등이 dangling 인지 확인하고, dangling 이면 메인 경로로 재바인딩.

**장점**: 기존 피해자 자동 구제.
**단점**: 매 셸 시작에 stat 비용. A/B 가 미래 발생을 막는 동안 C 는 한 번만 청소하면 됨 → migration helper 로 분리해도 됨.

권장 조합: **A (재발 방지) + C (1회 청소)**. C 는 `claude-accounts repair` 같은 명시적 명령으로 노출.

## 영향 범위 (Impact)

- 영향 받는 사용자: ai-worktree-spawn 을 적극 활용하는 사용자 전원 (= 본 repo 의 주 사용 시나리오)
- 우선순위: **중간~높음**. silent regression 이라 \"왜 안 되지\" 디버그 비용이 큼.
- 회피책 (현재): 메인 워크트리에서 `./setup.sh` 재실행 + 영향받은 `~/.claude-*/settings.json` 류 dangling 심볼릭 수동 정리. 즉시 임시 복구는:
  ```sh
  cd ~/dotfiles
  for d in ~/.claude ~/.claude-personal ~/.claude-work ~/.claude-work1; do
      [ -d \"$d\" ] || continue
      for f in settings.json statusline-command.sh; do
          if [ -L \"$d/$f\" ] && [ ! -e \"$d/$f\" ]; then
              rm \"$d/$f\"
              ln -s \"$HOME/dotfiles/claude/$f\" \"$d/$f\"
          fi
      done
  done
  ```

## 검증 (Verification)

1. **Unit (bats)** — `$PWD` 가 워크트리일 때 `_resolve_dotfiles_root` 가 메인 경로를 반환.
2. **Integration** —
   - 워크트리 생성 → 그 안에서 `claude_accounts_init` 호출 → `readlink ~/.claude-<acct>/settings.json` 이 메인 경로를 가리킴.
   - 워크트리 삭제 → 심볼릭 여전히 유효 (메인 경로 향함).
3. **Regression** — 메인에서 `./setup.sh` 호출은 기존과 동일하게 동작 (no-op).
4. **Manual** — 본 이슈의 \"재현\" 절차를 따라가서, 워크트리 삭제 후 `~/.claude-*/settings.json` 이 valid 임을 확인.

## 참고 (References)

- 관련 파일:
  - `shell-common/tools/integrations/claude.sh:1027-1049` — `_claude_account_setup_one`, symlink 생성 진입점
  - `shell-common/tools/integrations/claude.sh:862` — `_claude_dir_sync_one`, skills/docs 동기화
  - `shell-common/tools/integrations/claude.sh:760-791` — `_claude_ensure_symlink`, 타겟 비교 정책
  - `claude/setup.sh` — 진입점
  - `shell-common/env/` — `$DOTFILES_ROOT` 평가 위치
- 관련 이슈/커밋:
  - #584 / 0afa397 — settings.json SSOT 통합. 이번 버그와 직접 인과는 없지만, dangling 심볼릭이 발견된 맥락.
  - #571 — internal/external 환경별 분기.
  - ai-worktree-spawn / ai-worktree-teardown skill — 본 버그의 주 유발 시나리오.
- 환경: WSL2 (Linux 6.6.87.2-microsoft-standard-WSL2), zsh, Claude Code 2.1.140

---
<details>
<summary>🤖 AI Metrics · 📊 ~4500 tokens · 👤 ~2 h · 🤖 ~1 min</summary>

📊 ~4500 tokens · 👤 ~2 h · 🤖 ~1 min

</details>

### PR DESCRIPTION
**Pull request by dEitY719**
## Summary
- `${DOTFILES_ROOT}`가 워크트리 경로로 평가될 때 `~/.claude-*/`의 5종 심볼릭이 워크트리 경로로 박혀 워크트리 삭제 후 dangling → Claude Code가 default 로 fallback (statusline/hooks/플러그인 미동작) 되는 silent regression 수정.
- 메인 워크트리 정규화 헬퍼(`_resolve_dotfiles_root_canonical`)를 4개 진입점(`bash/main.bash`, `zsh/main.zsh`, `setup.sh`, `claude/setup.sh`)에 wiring — 향후 재발 방지.
- 이미 dangling 된 PC를 위한 1회 복구 명령 `claude-accounts repair [--dry-run]` 신설.

## Changes
- `shell-common/functions/dotfiles_root.sh` — `git rev-parse --git-common-dir` 로 워크트리 → 메인 워크트리 경로 환원. 순수 POSIX, fallback-safe. Escape hatch: `DOTFILES_ROOT_NO_CANONICALIZE=1`.
- `bash/main.bash` / `zsh/main.zsh` — 초기 `DOTFILES_ROOT` 계산 직후 헬퍼 source + `_dotfiles_root_canonicalize` 호출.
- `setup.sh` (top-level) — 워크트리에서 호출되면 메인 워크트리로 `cd` 한 뒤 후속 sub-script 호출. stderr 에 전환 사실을 출력해 사용자가 인지 가능.
- `claude/setup.sh` — defense-in-depth, 직접 호출 케이스(`./claude/setup.sh`) 도 보호.
- `shell-common/tools/integrations/claude.sh` — `claude_accounts_repair` 추가 (`~/.claude-*/` 의 5종 잘 알려진 심볼릭만 대상; 멱등; `--dry-run` 지원). `claude_accounts` dispatcher 와 help 에 `repair` 등록.
- `tests/bats/test_helper.bash` — `run_in_bash` / `run_in_zsh` 에 `DOTFILES_ROOT_NO_CANONICALIZE=1` 추가. 그렇지 않으면 워크트리에서 실행되는 bats 가 설치본(`~/dotfiles`) 코드를 검증하게 되어 PR 코드를 못 본다.
- 신규 bats 테스트: `tests/bats/functions/dotfiles_root.bats` (9 cases — 정규화 헬퍼 동작/escape hatch/edge case), `tests/bats/integrations/claude_accounts_repair.bats` (8 cases — 복구 함수 + dry-run + 멱등성).

## Test plan
- [x] 신규 bats 17/17 통과 (helper + repair).
- [x] 전체 bats suite 533/533 통과 — 회귀 없음.
- [x] `shellcheck` clean: 수정/신규 셸 스크립트 전부.
- [x] `shfmt -d -i 4 bash/main.bash` clean (tox 의 shfmt 스코프).
- [x] 수동 smoke: `/tmp/dr-smoke/{main,wt}` 임시 worktree 로 헬퍼 5종 케이스 검증.
- [ ] (post-merge) Internal/External PC 양쪽에서 `git worktree add ../wt-test some-branch && cd ../wt-test && ./setup.sh` 시 stderr "메인 워크트리로 전환" 메시지 + `~/.claude-*/settings.json` 메인 경로 확인.

## Related
Closes #589

---
<details>
<summary>🤖 AI Metrics · 📊 ~8500 tokens · 👤 ~2 h · 🤖 ~1 min</summary>

📊 ~8500 tokens · 👤 ~2 h · 🤖 ~1 min

</details>

### CODE REVIEW

File: shell-common/functions/dotfiles_root.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    _rdrc_common=$(git -C "$_rdrc_candidate" rev-parse --git-common-dir 2>/dev/null) || {
+        echo "$_rdrc_candidate"
+        return 0
+    }
+
+    if [ -z "$_rdrc_common" ]; then
+        echo "$_rdrc_candidate"
+        return 0
+    fi
```
[the remark this answers is not in the document]
Comment (dEitY719) [reply shell-common/functions/dotfiles_root.sh:83]: Accepted. Fixed in 91e0e7b — compare `--git-dir` with `--git-common-dir` and short-circuit when they are equal (covers both main repo and submodule checkouts). Verified the regression with a /tmp submodule fixture and added a bats case at `tests/bats/functions/dotfiles_root.bats` so this won't drift back.

**Comment by dEitY719** [comment]
> ## Code Review
>
> This pull request introduces a mechanism to canonicalize DOTFILES_ROOT to the main Git worktree, preventing dangling symlinks when linked worktrees are removed. It adds a new helper function in shell-common/functions/dotfiles_root.sh, integrates it into shell loaders and setup scripts, and provides a repair command in claude-accounts to fix existing tainted symlinks. Feedback suggests improving the canonicalization logic to correctly handle submodules by comparing the Git directory and common directory paths.

Accepted. The submodule edge case is now guarded by the git-dir / git-common-dir equality check in 91e0e7b, with a dedicated bats regression case so the path can't silently regress.

**Comment by dEitY719** [comment]
---
<details>
<summary>🤖 AI Metrics · 📊 ~3000 tokens · 👤 ~.50 h · 🤖 ~2 min</summary>


📊 ~3000 tokens · 👤 ~.50 h · 🤖 ~2 min


</details>
리뷰 답변: ~2 min · 사람: ~.50 h (2 comments × 0.25 h)
